doc/hardware/porting/board_porting.rst

```diff
@@ -242,6 +242,24 @@ followed by trial and error. If you want to understand details, you will need
 to read the rest of the devicetree documentation and the devicetree
 specification.
 
+Board extensions
+****************
+
+Board extensions are board fragments that may be present in any board root
+folder, under ``${BOARD_ROOT}/extensions/${BOARD}``. The board extension
+directory may contain Kconfig fragments and/or devicetree overlays. Board
+extensions are automatically loaded and applied on top of board files, before
+anything else. There is no guarantee on which order extensions are applied, in
```

Comment:
This is not quite clear to me: "Board extensions are automatically loaded and applied on top of board files, before
anything else."
For example, for my `nxp_adsp_imx8m`, I have `x.dts` and a `y.overlay`
When compiling:
- `samples/hello_world/` I need the `y.overlay`
- `../modules/audio/sof/app` I **_don't_** need the overlay

Therefore, my question is - will the` y.overlay` will be added, by default, independent of my sample?

Reply:
Board extensions, if present, are always applied, just on top of the board dts (before sample overlays, snippets, shields, etc.). It's a mechanism to extend a board from a module like example-application.

Reply:
while I see good reasons for this, I also share a concern similar to @iuliana-prodan .

This means that any Zephyr module can suddenly decide to add such extensions which will impact building of upstream samples, such as blinky and hello_world.

and if such extensions are not applied in a deterministic way, then it's not even possible to create a new board extension folder which overrules the other extension.

Another concern is that if an extension is applied which sets a config value then it's no longer possible to get it back to `unset`.
As example, sometimes samples / users just wants the `default y/n if <something>`, and not being set explicitly in most cases.
But if the board extension suddenly forces the value to a fixed `y`, then it's no longer possible to get it back to `unset`.
Users can force it to `n`, but that may not always be desirable, cause as mentioned above you might actually want it to be `y` in some cases.

So this behavior needs to be carefully thought through.
Also it must be guaranteed in which order such extensions are applied, so that users will know how to overrule when needed.

Reply:
> Board extensions, if present, are always applied, just on top of the board dts (before sample overlays, snippets, shields, etc.). It's a mechanism to extend a board from a module like example-application. Note it's independent of e.g. sample overlays.

Well, it's clear now, but I still don't think it's a good idea - to always apply it on top of dts.

For my above example, you're right, I can add my` y.overlay` in the sample folder. But I need this overlay in `hello_world`, `synchronization`, `philosophers`, etc., so I duplicate it everywhere I need it?

I agree with @tejlmand this behavior needs to be carefully thought through.

Reply:
> > Board extensions, if present, are always applied, just on top of the board dts (before sample overlays, snippets, shields, etc.). It's a mechanism to extend a board from a module like example-application. Note it's independent of e.g. sample overlays.
> 
> Well, it's clear now, but I still don't think it's a good idea - to always apply it on top of dts.
> 

So how you would add, for example, new hardware description?

> For my above example, you're right, I can add my` y.overlay` in the sample folder. But I need this overlay in `hello_world`, `synchronization`, `philosophers`, etc., so I duplicate it everywhere I need it?
> 

Yes, right now you need to duplicate it everywhere. Overlays are also not always an option, specially if you also have out-of-tree shields that depend on board definitions.

> I agree with @tejlmand this behavior needs to be carefully thought through.

Reply:
> In an overlay file given at compile time using -DDTC_OVERLAY_FILE

This is not always an option, because of the order on how overlays are applied. See the linked issue for more details. Also, this requires extra build step arguments.

Reply:
Why does an overlay imply bringing a HAL?

Btw, a real usecase of board extensions is shown here, so that the purpose of this PR can be seen in practice: https://github.com/teslabs/spinner/tree/board-refactor

Reply:
> Why does an overlay imply bringing a HAL?
> 
Because the driver implementation for a hw peripheral is in hal.

> Btw, a real usecase of board extensions is shown here, so that the purpose of this PR can be seen in practice: https://github.com/teslabs/spinner/tree/board-refactor
Thanks, I'll check it out

Reply:
> > Why does an overlay imply bringing a HAL?
> 
> Because the driver implementation for a hw peripheral is in hal.

I don't see how the current PR relates to that. If an overlay sets a certain peripheral to `status = "okay"`, nothing will happen unless the corresponding driver class is enabled. Also, considering this is a board extension, the same rules as upstream should prevail: enable minimal peripherals. But still, HALs are orthogonal to this discussion, this is proposing a mechanism to extend boards in oot aplications, like `example-application`.

Reply:
>  I just explained a case where an overlay, which enables a hw peripheral, doesn't have to be always applied over board dts.
> This is needed only for some samples and, in my opinion, too many to duplicate the overlay in each sample folder.

I think you're mixing this up with snippets. Snippet support was added recently and does what you're asking about. This addition is unrelated to snippets.